src/environment.rs

```diff
@@ -264,9 +302,97 @@ impl Environment {
     ///   processes need to either re-open the environment, or call set_map_size
     ///   with size 0 to update the environment. Otherwise, new transaction creation
     ///   will fail with `Error::MapResized`.
-    pub fn set_map_size(&mut self, size: size_t) -> Result<()> {
+    pub fn set_map_size(&self, size: size_t) -> Result<()> {
         unsafe { lmdb_result(ffi::mdb_env_set_mapsize(self.env(), size)) }
     }
+
+    // size_used doesn't include data yet to be committed. This will work only
+    // at the beginning of a transaction
+    fn map_occupied_size_inner(env_info: &Info, stat: &Stat) -> usize {
+        let size_used = stat.page_size() as usize * env_info.last_pgno();
+        size_used
+    }
+
+    /// Check whether a resize is needed under two conditions:
+    /// 1. The headroom + currently used size don't fit in map_size()
+    /// 2. More than RESIZE_PERCENTage is used of the database
+    fn needs_resize(&self, headroom: Option<usize>) -> Result<bool> {
+        let stat = self.stat()?;
+        let env_info = self.info()?;
+
+        let size_used = Self::map_occupied_size_inner(&env_info, &stat);
+
+        let current_map_size = env_info.map_size();
+
+        let current_percentage_used = size_used as f32 / current_map_size as f32;
+
+        if let Some(given_headroom) = headroom {
+            if env_info.map_size() < given_headroom.checked_add(size_used).expect("LMDB size check addition failed") {
+                return Ok(true);
+            }
+        }
+
+        let resize_settings = self.resize_settings.as_ref().unwrap_or(&DEFAULT_RESIZE_SETTINGS);
+
+        Ok(current_percentage_used > resize_settings.resize_trigger_percentage)
+    }
+
+    /// Do the resizing. This will pause all transactions (or will dead-lock), then resize,
+    /// and return the new map size.
+    /// Keep in mind that a single resize step cannot be larger than 1 << 31, due to usize limitations
+    /// this is due to the FFI using usize while lmdb uses mdb_size_t, which is always u64
+    pub fn do_resize(&self, increase_size: Option<usize>) -> Result<usize> {
+        let stat = self.stat()?;
+        let env_info = self.info()?;
+
+        let old_map_size = env_info.map_size();
+
+        let resize_settings = self.resize_settings.as_ref().unwrap_or(&DEFAULT_RESIZE_SETTINGS);
+        let increase_size = increase_size
+            .unwrap_or_else(|| Self::headroom_from_ratio(old_map_size, resize_settings.default_resize_ratio));
+        let increase_size = increase_size.clamp(resize_settings.min_resize_step, resize_settings.min_resize_step);
+
+        let current_occupied_ratio = Self::map_occupied_size_inner(&env_info, &stat);
+
+        // calculate new map size, and ensure it's an integer of OS page size
+        let new_map_size = old_map_size.checked_add(increase_size).expect("LMDB resize size addition failed");
+        let new_map_size = new_map_size + new_map_size % stat.page_size() as usize;
```

Comment:
Is this to align the map size to the page size boundary? If so, I don't think the code achieves that.

Reply:
I'm adding some asserts, but I can't check that the final size is old + headroom, because some times it has to be done iteratively.